data_structures/singly_linked_list.js

```diff
@@ -40,27 +48,126 @@ class Node {
  */
 
 class SinglyLinkedList {
-    constructor(){}
+    constructor(){
+        this.head = null
+        this.tail = null 
+    }
     
     //add a new node with the given value to the tail of the list
     add_to_tail(value){
-        
+        const new_node = new Node(value)
+
+        if(!this.head){
+            this.head = new_node
+            this.tail = new_node
+        }
+        else {
+            this.tail.next = new_node
+            this.tail = new_node
+        }
     }
     
     //returns the value of the removed node or null if list is empty
-    remove_tail(){}
+    remove_tail(){
+        if(!this.tail){
+            return null
+        }
+
+        let value = this.tail.value
+        let head = this.head
+
+        if(this.tail === this.head){
+            this.head = null
+            this.tail = null
+            return value
+        }
+
+        while(this.head.next.next !== null){
```

Comment:
You could use the node function get_next() here
```
while(current_node.get_next() !== this.tail)
```

Reply:
Instead of changing the list's head reference here and putting it back at the end you could create a current node to travel the list instead